src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+    }
+
+    /// <inheritdoc cref="SolutionState.RemoveProjectReference"/>
+    public SolutionCompilationState RemoveProjectReference(
+        (SolutionState newSolutionState, ProjectState oldProject, ProjectState newProject) tuple)
+    {
+        return ForkProject(
+            tuple,
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectReferences"/>
+    public SolutionCompilationState WithProjectReferences(
+        ProjectId projectId, IReadOnlyList<ProjectReference> projectReferences)
+    {
+        return ForkProject(
+            this.Solution.WithProjectReferences(projectId, projectReferences),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.AddMetadataReferences"/>
+    public SolutionCompilationState AddMetadataReferences(
+        ProjectId projectId, IReadOnlyCollection<MetadataReference> metadataReferences)
+    {
+        return ForkProject(
+            this.Solution.AddMetadataReferences(projectId, metadataReferences),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.RemoveMetadataReference"/>
+    public SolutionCompilationState RemoveMetadataReference(
+        (SolutionState newSolutionState, ProjectState oldProject, ProjectState newProject) tuple)
+    {
+        return ForkProject(
+            tuple,
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectMetadataReferences"/>
+    public SolutionCompilationState WithProjectMetadataReferences(
+        ProjectId projectId, IReadOnlyList<MetadataReference> metadataReferences)
+    {
+        return ForkProject(
+            this.Solution.WithProjectMetadataReferences(projectId, metadataReferences),
+            translate: null,
+            forkTracker: true);
+    }
+
+    /// <inheritdoc cref="SolutionState.AddAnalyzerReferences(ProjectId, ImmutableArray{AnalyzerReference})"/>
+    public SolutionCompilationState AddAnalyzerReferences(
+        (SolutionState newSolutionState, ProjectState oldProject, ProjectState newProject) tuple, ImmutableArray<AnalyzerReference> analyzerReferences)
+    {
+        return ForkProject(
+            tuple,
+            new CompilationAndGeneratorDriverTranslationAction.AddOrRemoveAnalyzerReferencesAction(tuple.oldProject.Language, referencesToAdd: analyzerReferences),
+            forkTracker: true);
+    }
+
+    public SolutionCompilationState AddAnalyzerReferences(
+        SolutionState newSolutionState)
+    {
+        // Note: This is the codepath for adding analyzers from vsixes.  Importantly, we do not ever get SGs added
+        // from this codepath, and as such we do not need to update the compilation trackers.  The methods that add SGs
+        // all come from entrypoints that are specific to a particular project.
+        return Branch(newSolutionState);
```

Comment:
an interesting case where we're not forking (this is not project affecting).  So we literally just only branch.  Keeping everythign the same except the solution state.